### PR TITLE
FTPx initialization

### PR DESCRIPTION
**Pull request by dallan-keylogic**
## Summary/Motivation:
I was having problems initializing flash calculations in which a noncondensable phase was present. It turned out that no attempt was made to calculate the vapor fraction or component phase splits when noncondensable or nonvaporizable phases were present in the `FTPx.py` initialization. 

## Changes proposed in this PR:
I derived a method (closely related to ideal bubble and dew point calculations) to calculate the vapor fraction of an ideal mixture obeying Raoult's law through a 1D root finding problem. Mathematically, one can prove that the root finding algorithm converges so long as the system pressure is between the ideal dew and bubble pressures, all the saturation pressures, Henry's law constants, and mole fractions are nonnegative, and the mole fractions add up to 1.  I added a failsafe for if the system pressure does not fall into that range, and exceptions if "significantly" nonphysical values for component parameters are calculated, as well as a failsafe if ill-conditioning or "insignificant" unphysical values cause a convergence failure.

A new IDAES exception, the `UserModelError` is added to handle cases in which unphysical values computed from user-supplied methods prevent execution from continuing.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by radhakrishnatg — DISMISSED** [review]

**Review by andrewlee94 — DISMISSED** [review]

**Comment by andrewlee94** [comment]
@dallan-keylogic It looks like there are some test failures associated with the NGCC models - at least one of them is related to the changes in this PR. You might need to work with @eslickj to fix these.

**Comment by andrewlee94** [comment]
@eslickj Any idea what is happening in the failing integration tests (NGCC models)?

File: idaes/generic_models/properties/core/state_definitions/FTPx.py [hunk 1]
```diff
@@ -284,6 +287,162 @@ def define_display_vars_FTPx():
...
+             vl_comps,
+             henry_comps,
+             l_only_comps,
+             v_only_comps) = _valid_VL_component_list(b, pp)
+            
+            H = {j: value(get_method(b, "henry_component", j, l_phase)(
+                        b, l_phase, j, b.temperature))
+                        for j in henry_comps}
+            try:
+                psat = {j: value(get_method(b, "pressure_sat_comp", j)(
```
[the remark this answers is not in the document]
Comment (dallan-keylogic) [reply idaes/generic_models/properties/core/state_definitions/FTPx.py:340]: Actually, yes we still do, in order to calculate the phase split of each component.

File: idaes/generic_models/properties/core/state_definitions/FTPx.py [hunk 2]
```diff
@@ -340,10 +331,11 @@ def state_initialization(b):
              henry_comps,
              l_only_comps,
              v_only_comps) = _valid_VL_component_list(b, pp)
+            pp_VLE = pp
             
-            H = {j: value(get_method(b, "henry_component", j, l_phase)(
-                        b, l_phase, j, b.temperature))
-                        for j in henry_comps}
+            # H = {j: value(get_method(b, "henry_component", j, l_phase)(
```
Comment (andrewlee94) [inline idaes/generic_models/properties/core/state_definitions/FTPx.py:336]: It API for Henry's Law should have stabilized now, so we should look at adding all of this back in.

Core code is in `idaes/generic_models/properties/core/phase_equil/henry.py`
See lines 762-764 in `/idaes/power_generation/carbon_capture/mea_solvent_system/properties/MEA_solvent.py` for an example of the new format.

This lets the user specify the basis for Henry's Law and we currently have support for using either mole fraction or concentration and both "Henry's Constant" and "Henry Volatility".

File: idaes/generic_models/properties/core/phase_equil/henry.py [hunk 3]
```diff
@@ -31,14 +31,15 @@ class HenryType(Enum):
...
-    Any differnt forms can use values 101+, but will need ot add the custom
+    Any differnt forms can use values 101+, but will need to add the custom
     code in if branches where necessary
     """
     # TODO: Add more forms as needed
     Hcp = 1
     Kpc = 51
     Hxp = 2
     Kpx = 52
+    Dummy = 999 # To test error handling
```
[the remark this answers is not in the document]
Comment (dallan-keylogic) [reply idaes/generic_models/properties/core/phase_equil/henry.py:42]: Unfortunately, using a negative number is incompatible with the logic as written:
```
    if henry_def["type"].value <= 50:
        # H = c/P type
        h_press = h_conc/henry
    elif henry_def["type"].value <= 100:
        # K = P/c type
        h_press = h_conc*henry
    else:
        _raise_henry_type_error(henry_def['type'])
```
This logic can, of course, be revised, but it's questionable whether or not its worth it.

**Review by andrewlee94 — APPROVED** [review]

File: idaes/generic_models/properties/core/phase_equil/tests/test_henry.py [hunk 4]
```diff
@@ -145,3 +149,89 @@ def test_constant_H():
...
+    henry_types = [HenryType.Hcp, HenryType.Kpc, HenryType.Hxp, HenryType.Kpx]
+    expected_split = [3/(5*7),(3*7)/5,1/(5*7),7/5]
+    
+    for htype,split in zip(henry_types,expected_split):
+        config_dict["components"]["H2O"]["henry_component"]["Liq"]["type"] = htype
+        m = ConcreteModel()
+        m.params = GenericParameterBlock(default=config_dict)
+        
+        m.state = m.params.build_state_block([0])
+        m.state[0].mole_frac_phase_comp["Liq","H2O"].value = 2
```
Comment (radhakrishnatg) [inline idaes/generic_models/properties/core/phase_equil/tests/test_henry.py:228]: Are you intentionally setting the mole fraction value to 2? Are you expecting to see a warning or error message here?

Comment (dallan-keylogic) [reply idaes/generic_models/properties/core/phase_equil/tests/test_henry.py:228]: Yes I am. The values used in this method don't have to be realistic, and in this case, it's testing that the mole fraction _doesn't_ affect the calculated value. However, that reminds me that, at some point, that might cause a Pyomo error. I'll change it to 1/2.

**Review by radhakrishnatg — APPROVED** [review]

**Review by andrewlee94 — APPROVED** [review]

**Review by radhakrishnatg — APPROVED** [review]

**Pull request merged**